### PR TITLE
Support different gc_safepoint for different keyspace

### PR DESCRIPTION
**Pull request by ystaticy**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tiflash/issues/7697

Problem Summary:

### Check List

Tests 

- [X] Manual test (add detailed scripts or steps below)

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Server/Server.cpp [hunk 1]
```diff
@@ -269,6 +270,8 @@ std::string Server::getDefaultCorePath() const
     return getCanonicalPath(config().getString("path")) + "cores";
 }
 
+bool PDClientHelper::enable_safepoint_v2 = false;
```
Comment (hongyunyan) [inline dbms/src/Server/Server.cpp:273]: Could you please put it into PDClientHelper.cpp? It's a little stange to initialize it here

File: dbms/src/Storages/Transaction/PDTiKVClient.h [hunk 2]
```diff
@@ -69,9 +111,72 @@ struct PDClientHelper
...
+
+    static KeyspaceGCInfo get_ks_gc_sp(KeyspaceID keyspace_id)
+    {
+        std::shared_lock<std::shared_mutex> lock(ks_gc_sp_mutex);
+        return ks_gc_sp_map[keyspace_id];
+    }
+
+    static void remove_ks_gc_sp(KeyspaceID keyspace_id)
+    {
+        std::shared_lock<std::shared_mutex> lock(ks_gc_sp_mutex);
```
Comment (hongyunyan) [inline dbms/src/Storages/Transaction/PDTiKVClient.h:167]: it should be std::unique_lock here.

File: dbms/src/Storages/Transaction/PDTiKVClient.h [hunk 3]
```diff
@@ -69,9 +111,72 @@ struct PDClientHelper
...
+        KeyspaceGCInfo ks_gc_info;
```
Comment (hongyunyan) [inline dbms/src/Storages/Transaction/PDTiKVClient.h:168]: It seems not be used here, remove it.

File: dbms/src/Server/Server.cpp [hunk 4]
```diff
@@ -269,6 +270,8 @@ std::string Server::getDefaultCorePath() const
     return getCanonicalPath(config().getString("path")) + "cores";
 }
 
+//bool PDClientHelper::enable_safepoint_v2 = false;
```
Comment (hongyunyan) [inline dbms/src/Server/Server.cpp:273]: please remove it.

**Review by hongyunyan — APPROVED** [review]
rest is LGTM

File: dbms/src/Storages/Transaction/PDTiKVClient.h [hunk 5]
```diff
@@ -31,17 +32,57 @@
...
         const pingcap::pd::ClientPtr & pd_client,
+        KeyspaceID keyspace_id,
         bool ignore_cache = true,
         Int64 safe_point_update_interval_seconds = 30)
     {
+        // If keyspace id is `NullspaceID` it need to use safe point v1.
+        if (enable_safepoint_v2 && keyspace_id != NullspaceID)
+        {
+            auto gc_safe_point = getGCSafePointV2WithRetry(pd_client, keyspace_id, ignore_cache, safe_point_update_interval_seconds);
+            LOG_DEBUG(Logger::get(), "use safe point v2, keyspace_id={},gc_safe_point={}", keyspace_id, gc_safe_point);
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Transaction/PDTiKVClient.h:82]:
```suggestion
            LOG_DEBUG(Logger::get(), "use safe point v2, keyspace={} gc_safe_point={}", keyspace_id, gc_safe_point);
```

Other codes use 'keyspace=xxx' in logging, so this is better for searching logs
https://github.com/pingcap/tiflash/blob/4ecf3cd7649f0597c540cdbfe4ea609f26d7d169/dbms/src/Storages/GCManager.cpp#L69

**Review by JaySon-Huang — APPROVED** [review]
LGTM with a logging style comment

**Pull request merged**